### PR TITLE
Remove Tensor(const Tensor&, BaseContext*, type)

### PR DESCRIPTION
**Pull request by dzhulgakov**
Differential Revision: D11915764

### CODE REVIEW

**Comment by bddppq** [comment]
rocm tests:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/3739/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/3728/

File: aten/src/ATen/core/TensorImpl.h [hunk 1]
```diff
@@ -847,6 +847,8 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
    * preserves the DeviceType of the source tensor (so, e.g., if you allocate
    * a tensor on CPU and then CopyFrom a CUDA tensor, that will to a
    * CUDA-to-CPU transfer).
+   *
+   * If the function is invoked without `context` the copy would be synchronous
```
Comment (ezyang) [inline aten/src/ATen/core/TensorImpl.h]: I'm actually not sure how this follows from the definition of this method below. Don't you need `FinishDeviceComputation()` to synchronize the copy? I don't see this method call in the body below.

**Review by ezyang — APPROVED** [review]
Approving because I assume you know what you're doing, but I would be quite interested to know why the sync call is no longer necessary.

File: aten/src/ATen/core/TensorImpl.h (hunk 1 above)
Comment (dzhulgakov) [reply aten/src/ATen/core/TensorImpl.h]: Ah! The destructor of Context object calls FinishDeviceComputation. Context is basically RAII if it's created on the fly :)

**Comment by dzhulgakov** [comment]
@ezyang - it basically replaces two  patterns of usage:
1) CPUContext only, where passing or not passing the context is the same (and it's cleaner to not pass)
2) CPU->GPU copy where people actually wanted to do a sync copy just after and invoke context.FinishDeviceComputation explicitly.

And yes, current logic is convoluted - that's why I'm untangling it

File: aten/src/ATen/core/TensorImpl.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/TensorImpl.h]: Great, thank you :)

**Pull request closed without merging**